### ISSUE TITLE
Reset to OS theme / colorMode should possible

### ISSUE THREAD

**Issue by juliusmarminge** [facebook/docusaurus#8074]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [X] I have tried the `npm run clear` or `yarn clear` command.
- [X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [X] I have tried creating a repro with https://new.docusaurus.io.
- [X] I have read the console error message carefully (if applicable).

### Description

I don't think this is a bug as much as intended behavior, but I think that intention can be confusing so I'll post an issue anyways.

Docusaurus is currently handling themes both using local storage, and syncing the user preferences. This often leads to the theme going out of sync:

https://github.com/facebook/docusaurus/blob/cba8be01a3ae058a328004ba281d28c47ac0a25c/packages/docusaurus-theme-classic/src/index.ts#L42-L52

### Reproducible demo

https://preferred-theme-test.vercel.app

### Steps to reproduce

Lets say at day I toggle the theme on the website to use light mode. That would store a KV-pair `{ "theme": "light" }` in localstorage. I leave the page and when I come back at night my device is now using darkmode, so I'd probably want the page to load in dark mode. However, since the KV-pair is in localstorage, that takes precedence so it loads light mode instead.

### Expected behavior

I'd want the page to load in dark mode.

### Actual behavior

The page loads the KV pair from localstorage and uses light mode.

### Your environment

- Public source code: https://github.com/juliusmarminge/preferred-theme-test
- Public site URL: https://preferred-theme-test.vercel.app
- Docusaurus version used: 2.1.0
- Environment name and version (e.g. Chrome 89, Node.js 16.4): Safari, Chrome
- Operating system and version (e.g. Ubuntu 20.04.2 LTS): MacOS 12

### Self-service

- [ ] I'd be willing to fix this bug myself.

**Comment by juliusmarminge** [comment]
I can submit a PR for this if you agree with my standpoint, but since I think this is intended I won't do anything until I hear back from you

**Comment by Josh-Cena** [comment]
The UX is quite hard to design. How should the localStorage value be deemed stale? Or should we not persist that value at all?

**Comment by juliusmarminge** [comment]
> The UX is quite hard to design. How should the localStorage value be deemed stale? Or should we not persist that value at all?

I usually don't persist it at all tbf but there is definitely a compromise to be made. 

\+ If you don't use localstorage, the theme is always in sync with the user's device, 

\- However say the user is on dark mode but want the page in light, the theme would reset to dark on every pageload.

I'd take the payoff and assume that users who want light pages probably has light mode enabled on their device

**Comment by juliusmarminge** [comment]
On https://trpc.io we override this behavior by a bit hacky script implemented here https://github.com/trpc/trpc/pull/2664

**Comment by Josh-Cena** [comment]
I think refreshing definitely should not change the color mode—that would feel really shaky. The only question is whether we can invalidate the localStorage somehow in the following case:

1. Light mode is selected—either as default or as user choice
2. User's device switched from light mode to dark mode the next time the site is opened

I think to do that we have to persist the current system theme in local storage as well?

(Also, I use system-wide dark mode, but the dark mode on some sites is just horrible that I use light mode on those instead.)

**Comment by juliusmarminge** [comment]
> I think refreshing definitely should not change the color mode—that would feel really shaky. The only question is whether we can invalidate the localStorage somehow in the following case:
> 
> 1. Light mode is selected—either as default or as user choice
> 2. User's device switched from light mode to dark mode the next time the site is opened
> 
> I think to do that we have to persist the current system theme in local storage as well?
> 
> (Also, I use system-wide dark mode, but the dark mode on some sites is just horrible that I use light mode on those instead.)

That sounds like a good idea. We could store a blob

```json
{"theme": "light", "userPreferred": "dark"}
```

and check if the mediaMatch === userPreferred in some way

**Comment by juliusmarminge** [comment]
I could play around and file a PR if you like?

**Comment by Josh-Cena** [comment]
Ah, sure, feel free to!

**Comment by flying-sheep** [comment]
Some people won’t like it, but this is a tri-state piece of data. *Any* UI that allows switching between all three states is preferable to the current one which makes one of them inaccessible. Working is better than broken, and only once things work, UX starts to become a topic.

I built this little thing a while ago btw: https://flying-sheep.github.io/react-color-scheme-switch/

**Comment by flying-sheep** [comment]
That’s exactly the UX the Python community is converging on. I didn’t have time to expand on it when I wrote my last comment, but here’s basically me saying the exact same things, https://github.com/rust-lang/this-week-in-rust/issues/2274#issuecomment-1333723860, with some example implementations.

**Comment by nasso** [comment]
how about keeping the same icon button, but with a third state? clicking it cycles between "light"/"dark"/"auto" (we just need a new icon, could be just an "A" or some i18n-friendly icon like a magic wand). when set to "auto", just store `{ "theme": "auto" }` in `localStorage`.

that would be more than enough and is better than what we have currently imho

**Comment by slorber** [comment]
Yes we agree we need 3 states and an icon to represent the new 3rd state.

That's not really worth pursing this discussion, but rather working on a concrete implementation based on my comments here:
https://github.com/facebook/docusaurus/issues/8074#issuecomment-1342662517

**Comment by nasso** [comment]
I'd like to work on that!

**Comment by ilg-ul** [comment]
Any progress on adding an _OS theme_ mode?

**Comment by Zwyx** [comment]
Just wanted to revive this. I've made a [demo project](https://zwyx.github.io/proper-dark-theme/) showing what I suggested above, and explained the implementation in [this article](https://zwyx.dev/blog/proper-dark-theme).

![image](https://github.com/facebook/docusaurus/assets/29386932/45909f42-8b76-4cf9-8e9b-c71837e91d4b)

@slorber  I'd be keen to know your opinion on this way of doing it :wink:

**Comment by mahozad** [comment]
Disclaimer: I'm the author of the library.

Would you like to incorporate [this animated theme switch](https://github.com/mahozad/theme-switch)?
It supports system (automatic) theme as well.

Demo 1: https://mahozad.ir/theme-switch/
Demo 2: https://mahozad.ir/android-pie-chart/

It probably addresses the following related issues too:
  - https://github.com/facebook/docusaurus/issues/8938
  - https://github.com/facebook/docusaurus/issues/8074
  - https://github.com/facebook/docusaurus/issues/7199
